File: src/cpp/src/utils/no_metals/resonance.cpp
#include "molgr/utils/no_metals/resonance.h"

#include "molgr/pipeline/resonance.h"
#include "molgr/stages/clean.h"
#include "molgr/stages/eliminate.h"
#include "molgr/stages/preprocess.h"
#include "molgr/utils/conversions.h"
#include "molgr/utils/electrons.h"
#include "molgr/utils/no_metals/selection.h"
#include "molgr/utils/resonance.h"
#include "molgr/compat/openbabel_iter.h"

#include <algorithm>
#include <chrono>
#include <cstddef>
#include <map>
#include <memory>
#include <optional>
#include <set>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

namespace
{
    using RawStateKey = molgr::no_metals::resonance::RawStateKey;
    using ProcessedStateKey = molgr::no_metals::resonance::ProcessedStateKey;

    struct RawResonanceCandidate
    {
        std::size_t seed_index = 0;
        std::size_t resonance_index = 0;
        std::size_t raw_index = 0;
        std::shared_ptr<OpenBabel::OBMol> omol;
    };

    std::vector<molgr::state::ReconstructionState> DeduplicateStates(
        std::vector<molgr::state::ReconstructionState> states)
    {
        std::set<RawStateKey> seen;
        std::vector<molgr::state::ReconstructionState> unique;
        for (auto &state : states)
        {
            RawStateKey key{
                molgr::resonance::BuildResonanceStateKey(state.Mol()),
                state.given_charge,
            };
            if (seen.insert(std::move(key)).second)
            {
                unique.push_back(std::move(state));
            }
        }
        return unique;
    }

    template <typename Stage>
    void AppendOmolStageVariants(
        std::vector<molgr::state::ReconstructionState> &pool,
        const std::string &phase,
        Stage stage)
    {
        std::vector<molgr::state::ReconstructionState> additions;
        for (const auto &state : pool)
        {
            auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
            auto next_mol = std::make_shared<OpenBabel::OBMol>(
                molgr::utils::CloneMolTopologyOnly(state.Mol()));
            machine = machine.Branch(std::nullopt, std::move(next_mol));
            if (machine.RunOmolStage(phase, stage))
            {
                additions.push_back(machine.FreezeLike(state));
            }
        }
        for (auto &state : additions)
        {
            pool.push_back(std::move(state));
        }
        pool = DeduplicateStates(std::move(pool));
    }

    bool RegisterTraversalLabel(
        std::map<RawStateKey, std::vector<std::pair<int, int>>> &labels_by_state,
        const RawStateKey &state_key,
        const std::pair<int, int> &label)
    {
        auto &labels = labels_by_state[state_key];
        for (const auto &known : labels)
        {
            if (known.first <= label.first && known.second <= label.second)
            {
                return false;
            }
        }
        labels.erase(
            std::remove_if(
                labels.begin(),
                labels.end(),
                [&label](const auto &known)
                {
                    return label.first <= known.first && label.second <= known.second;
                }),
            labels.end());
        labels.push_back(label);
        return true;
    }

    std::vector<molgr::state::ReconstructionState> PrepareCandidate(
        RawResonanceCandidate &raw,
        const molgr::state::ReconstructionState &seed,
        std::set<ProcessedStateKey> &seen_processed_states,
        const molgr::config::MolGRConfig &config)
    {
        auto machine = molgr::state::OmolStateMachine::FromReconstructionState(seed);
        machine = machine.Branch(
            "branch_resonance_candidate",
            std::move(raw.omol));
        const int target_radical_electrons = machine.total_radical_electrons;
        machine.RunOmolChargeStage(
            "process_resonance_eliminate_1_3_dipole_postive",
            molgr::reconstruct::Eliminate13DipolePostive);
        machine.RunOmolChargeStage(
            "process_resonance_eliminate_possible_cp_like_radical_anion",
            molgr::reconstruct::EliminatePossibleCPLikeRadicalAnion,
            machine.total_radical_electrons);
        machine.RunOmolStage(
            "process_resonance_clean_possible_1_3_dipole",
            molgr::reconstruct::CleanPossible13Dipole,
            machine.given_charge,
            machine.total_radical_electrons);
        machine.RunOmolStage(
            "process_resonance_clean_neighbor_radicals",
            molgr::reconstruct::CleanNeighborRadicals,
            machine.given_charge,
            machine.total_radical_electrons);
        machine.RunOmolStage(
            "process_resonance_clean_1_4_radicals",
            molgr::reconstruct::Clean14Radicals,
            machine.given_charge,
            machine.total_radical_electrons);
        machine.RunOmolStage(
            "process_resonance_clean_1_6_radicals",
            molgr::reconstruct::Clean16Radicals,
            machine.given_charge,
            machine.total_radical_electrons);
        machine.RunOmolChargeStage(
            "process_resonance_eliminate_positive_charges_1",
            [target_radical_electrons](OpenBabel::OBMol &mol, int &charge)
            {
                return molgr::reconstruct::EliminatePositiveChargesWithTarget(
                    mol, charge, target_radical_electrons);
            });
        machine.RunOmolChargeStage(
            "process_resonance_eliminate_negative_charges",
            [target_radical_electrons](OpenBabel::OBMol &mol, int &charge)
            {
                if (charge == 0)
                {
                    int real_unpaired = 0;
                    FOR_ATOMS_OF_MOL(atom_iter, mol)
                    {
                        real_unpaired += molgr::utils::GetUnpairedElectronCount(*atom_iter);
                    }
                    if (real_unpaired <= target_radical_electrons)
                    {
                        return false;
                    }
                }
                return molgr::reconstruct::EliminateNegativeCharges(mol, charge);
            });
        machine.RunOmolChargeStage(
            "process_resonance_eliminate_positive_charges_2",
            [target_radical_electrons](OpenBabel::OBMol &mol, int &charge)
            {
                return molgr::reconstruct::EliminatePositiveChargesWithTarget(
                    mol, charge, target_radical_electrons);
            });
        machine.RunOmolStage(
            "process_resonance_clean_resonances",
            molgr::reconstruct::CleanResonances);
        machine.Annotate("full_resonance_normalization");

        std::vector<molgr::state::ReconstructionState> candidates;
        std::vector<int> unresolved_indices;
        auto &validated_mol = machine.EnsureUniqueMol();
        FOR_ATOMS_OF_MOL(atom_iter, validated_mol)
        {
            if (molgr::utils::HasUnresolvedTwoElectronCenter(*atom_iter))
            {
                unresolved_indices.push_back(atom_iter->GetIdx());
            }
        }
        ProcessedStateKey processed_key{
            molgr::resonance::BuildProcessedResonanceKey(validated_mol),
            machine.given_charge,
        };
        if (!seen_processed_states.insert(std::move(processed_key)).second)
        {
            return candidates;
        }
        if (!molgr::reconstruct::ValidateOmol(
                validated_mol,
                seed.total_charge,
                seed.total_radical_electrons))
        {
            return candidates;
        }
        if (!unresolved_indices.empty())
        {
            int triplet_center_count = 0;
            for (int atom_idx : unresolved_indices)
            {
                if (molgr::utils::GetUnpairedElectronCount(
                        *validated_mol.GetAtom(atom_idx)) == 2)
                {
                    ++triplet_center_count;
                }
            }
            machine.Annotate("resolve_unresolved_two_electron_centers_at_validation");
            machine.metadata["unresolved_two_electron_singlet_centers"] =
                static_cast<int>(unresolved_indices.size()) - triplet_center_count;
            machine.metadata["unresolved_two_electron_triplet_centers"] =
                triplet_center_count;
        }
        machine.Annotate("validate_no_metal_candidate");
        machine.metadata["resonance_seed_index"] =
            static_cast<int>(raw.seed_index);
        machine.metadata["resonance_index"] =
            static_cast<int>(raw.resonance_index);
        machine.metadata["resonance_raw_index"] =
            static_cast<int>(raw.raw_index);
        machine.metadata["resonance_normalization"] = "full_resonance_normalization";
        auto candidate = machine.FreezeLike(seed);
        try
        {
            molgr::no_metals::selection::ScoreReconstructionCandidate(candidate, config);
        }
        catch (const std::exception &)
        {
            return candidates;
        }
        candidates.push_back(std::move(candidate));
        return candidates;
    }
}

namespace molgr
{
    namespace no_metals
    {
        namespace resonance
        {
            std::vector<molgr::state::ReconstructionState> BuildResonanceSeedPool(
                std::vector<molgr::state::ReconstructionState> neighbor_seeds)
            {
                auto pool = DeduplicateStates(std::move(neighbor_seeds));
                AppendOmolStageVariants(
                    pool,
                    "relocate_carbene_radical_for_resonance",
                    reconstruct::CleanCarbeneNeighborUnsaturated);
                return pool;
            }

    // Electron bookkeeping: radical traversal moves only real unpaired electrons.
    // Active lone pairs and unresolved markers stay in state keys. Validation is
    // the sole step that resolves deferred centers against the global spin budget.
    std::vector<molgr::state::ReconstructionState> SearchResonanceCandidates(
                const std::vector<molgr::state::ReconstructionState> &states,
                const molgr::config::MolGRConfig &config,
                molgr::pipeline::perf::RunTimingReducer *timing_reducer,
                ResonanceSearchSession *session)
            {
                ResonanceSearchSession local_session;
                auto &search_session = session == nullptr ? local_session : *session;
                std::vector<RawResonanceCandidate> raw_candidates;
                const int resonance_max_depth = std::max(0, config.resonance.max_depth);
                const int max_discrepancy = std::max(
                    0,
                    config.resonance.limited_discrepancy_max_discrepancy);

                const auto walk_started = std::chrono::steady_clock::now();
                std::size_t pruned_expansion_count = 0;
                for (std::size_t seed_index = 0; seed_index < states.size(); ++seed_index)
                {
                    std::size_t resonance_index = 0;
                    const auto &state = states[seed_index];
                    reconstruct::WalkRadicalResonancesLimitedDiscrepancy(
                        state.Mol(),
                        resonance_max_depth,
                        [&](const reconstruct::ResonanceSearchNode &node) -> bool
                        {
                            RawStateKey raw_key{node.state_key, state.given_charge};
                            bool should_expand = true;
                            if (node.depth < resonance_max_depth)
                            {
                                should_expand = RegisterTraversalLabel(
                                    search_session.labels_by_state,
                                    raw_key,
                                    {node.depth, node.discrepancy});
                                if (!should_expand)
                                {
                                    ++pruned_expansion_count;
                                }
                            }
                            if (search_session.seen_raw_states.insert(raw_key).second)
                            {
                                raw_candidates.push_back(
                                    RawResonanceCandidate{
                                        seed_index,
                                        resonance_index,
                                        search_session.next_raw_index++,
                                        node.omol_owner,
                                    });
                            }
                            ++resonance_index;
                            return should_expand;
                        },
                        reconstruct::LimitedDiscrepancyTraversalConfig{max_discrepancy},
                        config);
                }
                if (timing_reducer != nullptr)
                {
                    const auto now = std::chrono::steady_clock::now();
                    timing_reducer->AddResonanceWalkMs(
                        std::chrono::duration<double, std::milli>(now - walk_started).count());
                    timing_reducer->AddResonanceRawCandidates(
                        static_cast<double>(raw_candidates.size()));
                    timing_reducer->AddResonancePrunedExpansions(
                        static_cast<double>(pruned_expansion_count));
                }

                std::vector<molgr::state::ReconstructionState> candidates;
                const auto prepare_started = std::chrono::steady_clock::now();
                for (auto &raw : raw_candidates)
                {
                    const auto &seed = states[raw.seed_index];
                    auto prepared_candidates = PrepareCandidate(
                        raw,
                        seed,
                        search_session.seen_processed_states,
                        config);
                    for (auto &candidate : prepared_candidates)
                    {
                        candidates.push_back(std::move(candidate));
                    }
                }
                if (timing_reducer != nullptr)
                {
                    const auto now = std::chrono::steady_clock::now();
                    timing_reducer->AddResonancePrepareMs(
                        std::chrono::duration<double, std::milli>(now - prepare_started).count());
                    timing_reducer->AddResonancePreparedCandidates(
                        static_cast<double>(raw_candidates.size()));
                    timing_reducer->AddResonanceValidCandidates(
                        static_cast<double>(candidates.size()));
                    timing_reducer->AddResonanceDedupCandidates(
                        static_cast<double>(candidates.size()));
                }
                return candidates;
            }
        }
    }
}

File: src/molgr/fallback/utils/no_metals/resonance.py
"""Resonance recovery helpers for no-metal fallback reconstruction."""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field
from typing import Callable, Optional, Tuple, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.config import CONFIG, MolGRConfig
from molgr.fallback.pipeline.resonance import ResonanceTraversalPolicy, walk_radical_resonances
from molgr.fallback.stages.clean import clean_carbene_neighbor_unsaturated
from molgr.fallback.stages.preprocess import validate_omol
from molgr.fallback.state import (
    TRACE_NODE_METADATA_KEY,
    OmolStateMachine,
    ReconstructionState,
)
from molgr.fallback.utils import resonance as resonance_utils
from molgr.fallback.utils.electrons import (
    get_unpaired_electron_count,
    has_unresolved_two_electron_center,
)

from .selection import _annotate_no_metal_candidate_topology, _score_reconstruction_candidate


_OmolStage = Callable[[pybel.Molecule], Tuple[pybel.Molecule, bool]]
_OmolChargeStage = Callable[
    [pybel.Molecule, int],
    Tuple[pybel.Molecule, int, bool],
]


def _default_resonance_traversal_policy(
    config: MolGRConfig | None = None,
) -> ResonanceTraversalPolicy:
    resolved_config = CONFIG if config is None else config
    max_discrepancy = max(0, int(resolved_config.resonance.limited_discrepancy_max_discrepancy))
    traversal_score = resolved_config.resonance.traversal_score
    if traversal_score == "uff_lite_gain":
        return resonance_utils.make_limited_discrepancy_uff_lite_gain_traversal_policy(
            max_discrepancy=max_discrepancy,
        )
    if traversal_score == "input_order":
        return resonance_utils.make_limited_discrepancy_input_order_traversal_policy(
            max_discrepancy=max_discrepancy,
        )
    raise ValueError(
        "Unsupported resonance traversal_score. Expected 'uff_lite_gain' or 'input_order'."
    )


def _resonance_max_depth(config: MolGRConfig | None = None) -> int:
    resolved_config = CONFIG if config is None else config
    return max(0, int(resolved_config.resonance.max_depth))


def _clone_omol(omol: pybel.Molecule) -> pybel.Molecule:
    return pybel.Molecule(ob.OBMol(cast(ob.OBMol, omol.OBMol)))


@dataclass
class _RawResonanceCandidate:
    seed: ReconstructionState
    seed_index: int
    resonance_index: int
    raw_index: int
    omol: pybel.Molecule


@dataclass
class _ResonanceSearchSession:
    seen_raw_states: set[tuple[object, int]] = field(default_factory=set)
    labels_by_state: dict[tuple[object, int], list[tuple[int, int]]] = field(default_factory=dict)
    seen_processed_states: set[tuple[str, int]] = field(default_factory=set)
    next_raw_index: int = 0


def _candidate_key(state: ReconstructionState) -> tuple[object, int]:
    return resonance_utils.build_resonance_state_key(state.omol), state.given_charge


def _deduplicate_states(states: Iterable[ReconstructionState]) -> list[ReconstructionState]:
    unique: dict[tuple[object, int], ReconstructionState] = {}
    for state in states:
        unique.setdefault(_candidate_key(state), state)
    return list(unique.values())


def build_resonance_seed_pool(
    neighbor_seeds: Sequence[ReconstructionState],
) -> list[ReconstructionState]:
    """Add optional transformations that explicitly preserve electronic state."""

    pool = _deduplicate_states(neighbor_seeds)
    stages = (
        ("relocate_carbene_radical_for_resonance", clean_carbene_neighbor_unsaturated, False),
    )
    for phase, stage, uses_charge_budget in stages:
        pool = expand_resonance_seed_pool_stage(
            pool,
            phase=phase,
            stage=stage,
            uses_charge_budget=uses_charge_budget,
        )
    return pool


def expand_resonance_seed_pool_stage(
    states: Sequence[ReconstructionState],
    *,
    phase: str,
    stage: _OmolStage | _OmolChargeStage,
    uses_charge_budget: bool,
) -> list[ReconstructionState]:
    """Retain every input seed and add the changed result of one optional stage."""

    pool = _deduplicate_states(states)
    additions: list[ReconstructionState] = []
    for state in pool:
        machine = OmolStateMachine.from_reconstruction_state(state).branch(
            None,
            omol=_clone_omol(state.omol),
        )
        if uses_charge_budget:
            hit = machine.run_omol_charge_stage(
                phase,
                cast(_OmolChargeStage, stage),
            )
        else:
            hit = machine.run_omol_stage(phase, cast(_OmolStage, stage))
        if hit:
            additions.append(machine.freeze_like(state))
    return _deduplicate_states((*pool, *additions))


def _register_traversal_label(
    labels_by_state: dict[tuple[object, int], list[tuple[int, int]]],
    state_key: tuple[object, int],
    label: tuple[int, int],
) -> bool:
    labels = labels_by_state.setdefault(state_key, [])
    if any(depth <= label[0] and discrepancy <= label[1] for depth, discrepancy in labels):
        return False
    labels[:] = [
        (depth, discrepancy)
        for depth, discrepancy in labels
        if not (label[0] <= depth and label[1] <= discrepancy)
    ]
    labels.append(label)
    return True


def search_resonance_candidates(
    states: Sequence[ReconstructionState],
    *,
    resonance_traversal_policy: Optional[ResonanceTraversalPolicy] = None,
    config: MolGRConfig | None = None,
    session: _ResonanceSearchSession | None = None,
    walk_stage: Callable[..., None] | None = None,
    full_normalization_stage: _OmolChargeStage | None = None,
    validation_stage: Callable[[pybel.Molecule, int, int], bool] | None = None,
    score_stage: Callable[..., float] | None = None,
    topology_annotation_stage: Callable[..., object] | None = None,
) -> list[ReconstructionState]:
    """Search radical resonance states, then resolve deferred centers at validation.

    Traversal moves only real unpaired electrons; active lone pairs and unresolved
    markers remain part of the deduplication key. Each normalized candidate keeps
    unresolved centers until ``validate_omol`` assigns enough centers as triplets
    to satisfy the global radical budget and assigns the remainder as singlets.
    Impossible budgets are rejected without partially resolving a candidate.
    """

    walk_stage = walk_radical_resonances if walk_stage is None else walk_stage
    full_normalization_stage = (
        resonance_utils.process_resonance
        if full_normalization_stage is None
        else full_normalization_stage
    )
    validation_stage = validate_omol if validation_stage is None else validation_stage
    score_stage = _score_reconstruction_candidate if score_stage is None else score_stage
    topology_annotation_stage = (
        _annotate_no_metal_candidate_topology
        if topology_annotation_stage is None
        else topology_annotation_stage
    )
    search_session = _ResonanceSearchSession() if session is None else session
    raw_candidates: list[_RawResonanceCandidate] = []
    use_dominance_pruning = resonance_traversal_policy is None or isinstance(
        resonance_traversal_policy,
        resonance_utils._LIMITED_DISCREPANCY_POLICY_TYPES,
    )
    resonance_max_depth = _resonance_max_depth(config)
    for seed_index, state in enumerate(states):
        resonance_index = 0

        def collect(
            node: resonance_utils.ResonanceSearchNode,
            state: ReconstructionState = state,
            seed_index: int = seed_index,
        ) -> bool:
            nonlocal resonance_index
            raw_key = (node.state_key, state.given_charge)
            should_expand = True
            if use_dominance_pruning and node.depth < resonance_max_depth:
                should_expand = _register_traversal_label(
                    search_session.labels_by_state,
                    raw_key,
                    (node.depth, node.discrepancy),
                )
            if raw_key not in search_session.seen_raw_states:
                search_session.seen_raw_states.add(raw_key)
                raw_candidates.append(
                    _RawResonanceCandidate(
                        seed=state,
                        seed_index=seed_index,
                        resonance_index=resonance_index,
                        raw_index=search_session.next_raw_index,
                        omol=node.omol,
                    )
                )
                search_session.next_raw_index += 1
            resonance_index += 1
            return should_expand

        walk_stage(
            state.omol,
            max_depth=resonance_max_depth,
            traversal_policy=resonance_traversal_policy,
            visit=collect,
        )

    candidates: list[ReconstructionState] = []
    for raw in raw_candidates:
        machine = OmolStateMachine.from_reconstruction_state(raw.seed).branch(
            "branch_resonance_candidate",
            omol=_clone_omol(raw.omol),
            metadata={
                "resonance_seed_index": raw.seed_index,
                "resonance_index": raw.resonance_index,
                "resonance_raw_index": raw.raw_index,
                "resonance_normalization": "full_resonance_normalization",
            },
        )
        if full_normalization_stage is resonance_utils.process_resonance:
            resonance_utils._run_process_resonance(machine, emit_summary=True)
        else:
            machine.run_omol_charge_stage(
                "full_resonance_normalization",
                full_normalization_stage,
            )
        unresolved_indices = [
            atom.idx
            for atom in machine.omol.atoms
            if has_unresolved_two_electron_center(atom.OBAtom)
        ]
        processed_key = (
            resonance_utils.build_processed_resonance_key(machine.omol),
            machine.given_charge,
        )
        if processed_key in search_session.seen_processed_states:
            machine.trace_checkpoint(
                "discard_duplicate_processed_resonance_candidate",
                reason="processed_state_already_seen",
            )
            continue
        search_session.seen_processed_states.add(processed_key)
        if not validation_stage(
            machine.omol,
            raw.seed.total_charge,
            raw.seed.total_radical_electrons,
        ):
            machine.trace_checkpoint(
                "reject_no_metal_candidate_validation",
                reason="charge_or_radical_target_mismatch",
            )
            continue
        if unresolved_indices:
            triplet_center_count = sum(
                get_unpaired_electron_count(machine.omol.OBMol.GetAtom(atom_idx)) == 2
                for atom_idx in unresolved_indices
            )
            machine.annotate(
                "resolve_unresolved_two_electron_centers_at_validation",
                unresolved_two_electron_singlet_centers=(
                    len(unresolved_indices) - triplet_center_count
                ),
                unresolved_two_electron_triplet_centers=triplet_center_count,
            )
        machine.annotate(
            "validate_no_metal_candidate",
            resonance_seed_index=raw.seed_index,
            resonance_index=raw.resonance_index,
            resonance_raw_index=raw.raw_index,
            resonance_normalization="full_resonance_normalization",
        )
        candidate = machine.freeze_like(raw.seed)
        try:
            score_stage(candidate, config=config)
        except ValueError:
            machine.trace_checkpoint(
                "reject_no_metal_candidate_scoring",
                reason="force_field_scoring_failed",
            )
            continue
        topology_annotation_stage(candidate, config=config)
        machine.trace_checkpoint(
            "accept_no_metal_candidate",
            score=candidate.metadata.get("score"),
        )
        if TRACE_NODE_METADATA_KEY in machine.metadata:
            candidate.metadata[TRACE_NODE_METADATA_KEY] = machine.metadata[TRACE_NODE_METADATA_KEY]
        candidates.append(candidate)

    return candidates


__all__ = [
    "ResonanceTraversalPolicy",
    "_default_resonance_traversal_policy",
    "build_resonance_seed_pool",
    "expand_resonance_seed_pool_stage",
    "search_resonance_candidates",
]
